### PR TITLE
feat: add full GCP container deployment

### LINKED ISSUE #108: GCP container deployment

### PR DESCRIPTION
**Pull request by Mearman**
- closes #108

### CODE REVIEW

File: .github/workflows/reusable-gcp-docker-build-push.yml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          workload_identity_provider: ${{ secrets.GCP_WORKLOAD_IDENTITY_PROVIDER }}
+          project_id: ${{ inputs.gcp_project_id }}
+
+      - name: Set up Cloud SDK
+        uses: google-github-actions/setup-gcloud@v2
+
+      - name: Configure Docker for GCR
+        run: gcloud auth configure-docker ${{ inputs.gcp_region }}-docker.pkg.dev
+
+      # - name: Login to Artifact Registry
```
Comment (wfaithfull) [inline .github/workflows/reusable-gcp-docker-build-push.yml:95]: Can we remove this if it's not needed?

File: .github/actions/docker-google-artifact-registry/action.yml
Comment (wfaithfull) [inline .github/actions/docker-google-artifact-registry/action.yml:1]: Let me just clarify my understanding on these files. This is a composite action that pushes a pre-built docker image to GCP/GAR, is that correct?

File: .github/workflows/gcp-docker.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+jobs:
+  base:
+    name: Base image
+    uses: ./.github/workflows/reusable-gcp-docker-build-push.yml
+    with:
+      image_name: base
+      dockerfile: ./Dockerfile
+      context: .
+      push: ${{ github.event.inputs.push || github.ref == 'refs/heads/main' }}
```
Comment (wfaithfull) [inline .github/workflows/gcp-docker.yml:30]: If these are all scoped to `refs/heads/main` individually could this not just be only triggered on the main branch?

File: .github/actions/docker-google-artifact-registry/action.yml
Comment (wfaithfull) [reply .github/actions/docker-google-artifact-registry/action.yml:1]: Where is this action actually used?

**Review by wfaithfull — COMMENTED** [review]
Just a few clarifying questions and observations, nice work.

File: .github/workflows/gcp-docker.yml (hunk 2 above)
Comment (Mearman) [reply .github/workflows/gcp-docker.yml:30]: yes though I'm actually using it to control the push specifically so the build itself can still be tested for non-`main` branches.
however, the `builder` and `runtime` could be disabled and just use the `monolith` as the "canary"?

File: .github/actions/docker-google-artifact-registry/action.yml
Comment (Mearman) [reply .github/actions/docker-google-artifact-registry/action.yml:1]: I was keeping this for cases where the reusable workflow may not be sufficient, but I think I'd likely just create a fresh workflow now

File: .github/workflows/reusable-gcp-docker-build-push.yml (hunk 1 above)
Comment (Mearman) [reply .github/workflows/reusable-gcp-docker-build-push.yml:95]: yup can be dropped now. I was keeping it as there were some scenarios `gcloud auth configure-docker` didn't work for but they seem to be covered now

Comment (Mearman) [reply .github/workflows/reusable-gcp-docker-build-push.yml:95]: yup can be dropped now. I was keeping it as there were some scenarios `gcloud auth configure-docker` didn't work for but they seem to be covered now

File: .github/workflows/reusable-gcp-docker-build-push.yml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      #   uses: docker/login-action@v3
+      #   with:
+      #     registry: ${{ inputs.gcp_region }}-docker.pkg.dev
+      #     username: oauth2accesstoken
+      #     password: ${{ steps.auth.outputs.access_token }}
+
```
Comment (Mearman) [inline .github/workflows/reusable-gcp-docker-build-push.yml:101]:
```suggestion

```

File: .github/workflows/gcp-docker.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      build_args: |
+        BASE_IMAGE=${{ vars.GAR_REGION }}-docker.pkg.dev/${{ vars.GCP_PROJECT_ID }}/${{ vars.GAR_REPOSITORY }}/base
+      gcp_project_id: ${{ vars.GCP_PROJECT_ID }}
+      repository_path: ${{ vars.GAR_REGION }}-docker.pkg.dev/${{ vars.GCP_PROJECT_ID }}/${{ vars.GAR_REPOSITORY }}
+    secrets:
+      GCP_SERVICE_ACCOUNT: ${{ secrets.GCP_SERVICE_ACCOUNT }}
+      GCP_WORKLOAD_IDENTITY_PROVIDER: ${{ secrets.GCP_WORKLOAD_IDENTITY_PROVIDER }}
+
+  runtime:
```
Comment (Mearman) [inline .github/workflows/gcp-docker.yml:57]:
```suggestion
  runtime:
      if: ${{ github.ref == 'refs/heads/main' }}
```

File: .github/workflows/gcp-docker.yml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    name: Runtime image
+    needs:
+      - base
+    uses: ./.github/workflows/reusable-gcp-docker-build-push.yml
+    with:
+      image_name: runtime
+      dockerfile: ./rust/Dockerfile
+      context: .
+      push: ${{ github.event.inputs.push || github.ref == 'refs/heads/main' }}
```
Comment (Mearman) [inline .github/workflows/gcp-docker.yml:66]:
```suggestion
      push: true
```

File: .github/workflows/gcp-docker.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        BINARY_PATH=target/release/runtime
+        FILES=target/release/runtime
+        EXPOSED_PORT=8082
+      gcp_project_id: ${{ vars.GCP_PROJECT_ID }}
+      repository_path: ${{ vars.GAR_REGION }}-docker.pkg.dev/${{ vars.GCP_PROJECT_ID }}/${{ vars.GAR_REPOSITORY }}
+    secrets:
+      GCP_SERVICE_ACCOUNT: ${{ secrets.GCP_SERVICE_ACCOUNT }}
+      GCP_WORKLOAD_IDENTITY_PROVIDER: ${{ secrets.GCP_WORKLOAD_IDENTITY_PROVIDER }}
+
+  builder:
```
Comment (Mearman) [inline .github/workflows/gcp-docker.yml:79]:
```suggestion
  builder:
      if: ${{ github.ref == 'refs/heads/main' }}
```

File: .github/workflows/gcp-docker.yml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    name: Builder image
+    needs:
+      - base
+    uses: ./.github/workflows/reusable-gcp-docker-build-push.yml
+    with:
+      image_name: builder
+      dockerfile: ./rust/Dockerfile
+      context: .
+      push: ${{ github.event.inputs.push || github.ref == 'refs/heads/main' }}
```
Comment (Mearman) [inline .github/workflows/gcp-docker.yml:88]:
```suggestion
      push: true
```

File: .github/workflows/gcp-docker.yml (hunk 2 above)
Comment (Mearman) [reply .github/workflows/gcp-docker.yml:30]: see comments below

Comment (Mearman) [reply .github/workflows/gcp-docker.yml]: @wfaithfull updated pipeline running now if you want to look?

Comment (wfaithfull) [reply .github/workflows/gcp-docker.yml]: Ah fantastic, yes, I see. It is skipping the pushes

**Review by wfaithfull — APPROVED** [review]
LGTM, assuming the remainder of the pipeline passes.

**Pull request merged**